akka-typed/src/main/scala/akka/typed/internal/ActorCell.scala

```diff
@@ -315,10 +312,39 @@ private[typed] class ActorCell[T](
          * activation token again.
          */
         val again = unsafe.getAndAddInt(this, statusOffset, 1)
-        if (activations(again) == 0) executionContext.execute(this)
-        else unsafe.getAndAddInt(this, statusOffset, -1)
+        if (activations(again) == 0) true //reschedule
+        else {
+          unsafe.getAndAddInt(this, statusOffset, -1)
+          false // don't reschedule
+        }
+      } else {
+        false // don't reschedule
       }
     }
+
+    val reschedule =
+      try {
+        unscheduleReceiveTimeout()
+        if (!isTerminated(status)) {
+          process()
+          scheduleReceiveTimeout()
+        }
+        postProcess()
+      } catch {
+        case NonFatal(ex) ⇒
+          fail(ex)
+          postProcess()
+        case ie: InterruptedException ⇒
+          fail(ie)
+          if (Debug) println(s"[$thread] $self interrupting due to catching InterruptedException")
+          Thread.currentThread.interrupt()
+          postProcess()
+      }
```

Comment:
Looking at it this way, it seems to me that it would be cleaner to move `val reschedule = postProcess()` after the try–catch statement.

And we could think about doing `postProcess()` and re-throw for fatal exceptions, since `postProcess()` is not doing allocations and probably not even growing the stack (I don’t have a very strong opinion on this second part, though).